sidecar/internal/csiaddonsnode/csiaddonsnode.go

```diff
@@ -275,3 +281,76 @@ func generateName(nodeID, namespace, ownerKind, ownerName string) (string, error
 
 	return base, nil
 }
+
+// watchCSIAddonsNode starts a watcher for a specific CSIAddonsNode resource identified by its name.
+// If a CSIAddonsNode is deleted, it logs a warning and attempts to recreate it using mgr.Deploy()
+func (mgr *Manager) watchCSIAddonsNode(node *csiaddonsv1alpha1.CSIAddonsNode) error {
+	// Call deploy on start, this takes care of the cases where
+	// a watcher might exit due to an error while trying to
+	// recreate CSIAddonsNode in the cluster
+	err := mgr.Deploy(node)
+	if err != nil {
+		klog.Fatalf("Failed to create csiaddonsnode: %v", err)
+	}
+
+	klog.Infof("Starting watcher for CSIAddonsNode: %s", node.Name)
+
+	dynamicClient, err := dynamic.NewForConfig(mgr.Config)
+	if err != nil {
+		return fmt.Errorf("failed to create dynamic client: %w", err)
+	}
+
+	gvr := schema.GroupVersionResource{
+		Group:    csiaddonsv1alpha1.GroupVersion.Group,
+		Version:  csiaddonsv1alpha1.GroupVersion.Version,
+		Resource: "csiaddonsnodes",
+	}
+
+	watcher, err := dynamicClient.Resource(gvr).Namespace(mgr.PodNamespace).Watch(context.Background(), v1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", node.Name),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to watch CSIAddonsNode objects: %w", err)
+	}
+	defer watcher.Stop()
+
+	for event := range watcher.ResultChan() {
+		switch event.Type {
+		case watch.Deleted:
+			klog.Infof("WARNING: An active CSIAddonsNode: %s was deleted, it will be recreated", node.Name)
+
+			err := mgr.Deploy(node)
+			if err != nil {
+				return fmt.Errorf("failed to recreate CSIAddonsNode: %w", err)
+			}
+			klog.Infof("CSIAddonsNode: %s recreated", node.Name)
+		}
+	}
+
+	// The channel was closed by the API server without any errors
+	// Simply log it here and return, the dispatcher is responsible
+	// for restarting the watcher
+	klog.Infof("Watcher for %s exited gracefully, will be restarted soon", node.Name)
+
+	return nil
+}
+
+// DispatchWatcher starts a watcher for the specified CSIAddonsNode and retries
+// if the watcher exits due to an error. It will retry up to a maximum number of
+// attempts defined by watcherRetryCount before returning an error.
+func (mgr *Manager) DispatchWatcher(node *csiaddonsv1alpha1.CSIAddonsNode) error {
+	retryCount := 0
+
+	for retryCount < int(watcherRetryCount) {
+		err := mgr.watchCSIAddonsNode(node)
+		if err != nil {
+			klog.Errorf("Watcher for %s exited, retrying (%d/%d), error: %v", node.Name, retryCount+1, watcherRetryCount, err)
+
+			retryCount++
+			time.Sleep(watcherRetryDelay)
```

Comment:
can we make it like 5 sec, 10 sec 20 sec, keep 5 as default and multiple it by retryCount. we dont need to wait for 1 minute.